R/unused_import_linter.R

```diff
@@ -0,0 +1,64 @@
+#' Check that imported packages are actually used
+#'
+#' @param except_packages Character vector of packages that are ignored.
+#' These are usually attached for their side effects.
+#'
+#' @evalRd rd_tags("unused_import_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+unused_import_linter <- function(except_packages = c("bit64", "data.table", "tidyverse")) {
+  Linter(function(source_file) {
+    if (is.null(source_file$full_xml_parsed_content)) return(list())
+
+    import_exprs <- xml2::xml_find_all(
+      source_file$full_xml_parsed_content,
+      paste(
+        "//expr[",
+        "expr[SYMBOL_FUNCTION_CALL[text() = 'library' or text() = 'require']]",
+        "and",
+        "not(SYMBOL_SUB[text() = 'character.only'])",
+        "]"
+      )
+    )
+    if (length(import_exprs) == 0) {
+      return(list())
+    }
+    imported_pkgs <- xml2::xml_text(xml2::xml_find_first(import_exprs, "expr[STR_CONST|SYMBOL]"))
+    imported_pkgs <- vapply(
+      imported_pkgs,
+      function(expr) as.character(parse(text = expr, keep.source = FALSE)),
```

Comment:
can we vectorize this so we only have to run `parse()` once?

Reply:
Interesting enough, `as.character(parse(text = imported_pkgs, keep.source = FALSE))` seems to work in this context.
I interactively tested curious edge cases (extremely long strings, and the empty string).